tower-http/src/classify.rs

```diff
@@ -0,0 +1,313 @@
+    type Classifier = C;
+
+    fn make_classify<B>(&self, _req: &Request<B>) -> Self::Classifier
+    where
+        B: Body,
+    {
+        self.classifier.clone()
+    }
+}
+
+/// Trait for classifying responses as either success or failure. Designed to support both unary
+/// requests (single request for a single response) as well as streaming responses.
+///
+/// Can for example be used in logging or metrics middlewares, or to build [retry policies].
+///
+/// [retry policies]: https://docs.rs/tower/latest/tower/retry/trait.Policy.html
+pub trait ClassifyResponse<E> {
+    /// The type of failure classifications.
+    type FailureClass;
+
+    /// The type used to classify the response end of stream (EOS).
+    type ClassifyEos: ClassifyEos<FailureClass = Self::FailureClass>;
+
+    /// Classify a response.
+    ///
+    /// Returns a [`ClassifiedResponse`] which specifies whether the response was able to
+    /// classified immediately or if we have to wait until the end of a streaming response.
+    fn classify_response<B>(
+        self,
+        res: &Response<B>,
+    ) -> ClassifiedResponse<Self::FailureClass, Self::ClassifyEos>
+    where
+        B: Body;
+
+    /// Classify an error.
+    ///
+    /// Errors are always errors (doh) but sometimes it might be useful to have multiple classes of
+    /// errors. A retry policy might allows retrying some errors and not others.
+    fn classify_error(self, error: &E) -> Self::FailureClass;
+}
+
+/// Trait for classifying end of streams (EOS) as either success or failure.
+pub trait ClassifyEos {
+    /// The type of failure classifications.
+    type FailureClass;
+
+    /// Perform the classification from response trailers.
+    fn classify_eos(self, trailers: Option<&HeaderMap>) -> Result<(), Self::FailureClass>;
+}
+
+/// Result of doing a classification.
+#[derive(Debug)]
+pub enum ClassifiedResponse<FailureClass, ClassifyEos> {
+    /// The response was able to be classified immediately.
+    Ready(Result<(), FailureClass>),
+    /// We have to wait until the end of a streaming response to classify it.
+    RequiresEos(ClassifyEos),
+}
+
+/// A [`ClassifyEos`] type that can be used in [`ClassifyResponse`] implementations that never have
+/// to classify streaming responses.
+///
+/// `NeverClassifyEos` exists only as type. You cannot construct `NeverClassifyEos` values.
+pub struct NeverClassifyEos<T> {
+    _output_ty: PhantomData<fn() -> T>,
+    _never: Infallible,
+}
+
+impl<T> ClassifyEos for NeverClassifyEos<T> {
+    type FailureClass = T;
+
+    fn classify_eos(self, _trailers: Option<&HeaderMap>) -> Result<(), Self::FailureClass> {
+        // `NeverClassifyEos` contains an `Infallible` so it can never be constructed
+        unreachable!()
+    }
+}
+
+/// The default classifier used for normal HTTP responses.
+///
+/// Responses with a `5xx` status code are failures, all others as successes.
+#[derive(Clone, Debug, Default)]
+pub struct ServerErrorsAsFailures {
```

Comment:
Would one that considered client errors as well as server errors errors be useful? Similar to what reqwest does

https://docs.rs/reqwest/0.11.1/src/reqwest/async_impl/response.rs.html#341

Reply:
Yeah totally. I was thinking about adding one that took a list of ranges that would be considered errors. I think I'll do that in a follow up PR though.

Made an issue for tracking that https://github.com/tower-rs/tower-http/issues/73